### PR TITLE
Add ability to use the Cinder volume plugin w/o OpenStack provider

### PR DESCRIPTION
**Pull request by Quentin-M**
**What this PR does / why we need it**:

Introduced a few weeks back at SIG Storage, this PR adds the ability to the existing Cinder plugin to manage and provision OpenStack Cinder volumes without the need to use the OpenStack provider. Instead, users can simply specify a secret containing their own OpenStack Keystone credentials, along with a volume identifier.

While currently the plugin only supports the _rbd_ backend (any other Cinder backends can be implemented as necessary moving forward), it differs from the existing _rbd_ driver by abstracting away any technical details and sensitive configuration from the end-user, leaving that responsibility to the cluster administrator (i.e. _ceph.conf_, _ceph.client.admin.keyring_); Therefore respecting the separation of knowledge/privilege established between OpenStack administrators and users. The plugin also supports dynamic provisioning.

Below are examples/excerpts of manifests that use this plugin.

**Keystone credentials** 

```
apiVersion: v1
kind: Secret
metadata:
  name: cinder-auth
type: kubernetes.io/cinder
data:
  authURL: aHR0cDovL2lkZW50aXR5Lm9wZW5zdGFjay5jbHVzdGVyOjMwMDgwL3YzLw==
  username: YWRtaW4=
  password: YWRtaW4=
  project: YWRtaW4=
  domain: ZGVmYXVsdA==
```

(and optional `caCert`, `cert`/`key` fields to TLS authentication)

**Defining a Cinder volume for a pod mount**

```
  - name: cinder-volume
    cinder:
      volumeID: cdf62f61-3120-4946-b68d-5371894cc890
      secretRef: cinder-auth
      fsType: ext4
```

**StorageClass for dynamic provisioning**

```
kind: StorageClass
apiVersion: storage.k8s.io/v1beta1
metadata:
  name: cinder-volumes
provisioner: kubernetes.io/cinder
parameters:
  secretRef: cinder-auth
```

**Claiming a new Cinder volume for a pod mount**

```
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: cinder-claim
  annotations:
    volume.beta.kubernetes.io/storage-class: "cinder-volumes"
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
```

**Special notes for your reviewer**:

The PR has a huge lines count because the `gophercloud/gophercloud` dependency is introduced and because there're changes in the generated code. Of course, `rackspace/gophercloud` is already vendored in Kubernetes, but the project is currently in the state of [feature freeze](https://github.com/rackspace/gophercloud/issues/592). Unfortunately, we couldn't simply use `rackspace/gophercloud` in all good conscience because multi-attach support isn't there and because new features such as [BeginDetaching](https://github.com/gophercloud/gophercloud/pull/90) are theoretically needed. On the other hand, we couldn't replace entirely the `rackspace/gophercloud` dependency because the new repository doesn't contain the _rackspace/_ directory, currently used by the _rackspace_ cloud provider. The feature freeze issue mentions that the directory will be moved to its own repository. I am sure that the owners/maintainers of the _rackspace_ provider's (no file present in the directory) would be able to tell me more and may come-up with a follow-up PR.

Thank you for your time.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33944)

### CODE REVIEW

File: pkg/volume/cinder_local/cinder_util.go [hunk 1]
```diff
@@ -0,0 +1,573 @@
...
+	secret := &api.Secret{}
+	_, _, err = s.Decode(contents, nil, secret)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode persisted secret file: %v", err)
+	}
+
+	return secret, nil
+}
+
+func newCinderClient(host volume.VolumeHost, secret *api.Secret) (*gophercloud.ServiceClient, error) {
```
Comment (rootfs) [inline pkg/volume/cinder_local/cinder_util.go]: nit: please split the file into openstack util and cinder util

**Comment by rootfs** [comment]
@kubernetes/sig-openstack 

There is a request to migrate from rackspace to gophercloud #30404.

File: pkg/volume/cinder_local/cinder.go [hunk 2]
```diff
@@ -0,0 +1,432 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
Comment (Quentin-M) [inline pkg/volume/cinder_local/cinder.go]: Changed.

File: pkg/volume/cinder_local/cinder.go [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+	AttachDisk(mounter *cinderVolumeMounter, mntPoint string) error
+	// Detaches the disk from the kubelet's host machine.
+	DetachDisk(unmounter *cinderVolumeUnmounter, mntPoint string) error
+	// Creates a volume
+	CreateVolume(provisioner *cinderVolumeProvisioner) (volumeID string, volumeSizeGB int, secretRef string, err error)
+	// Deletes a volume
+	DeleteVolume(deleter *cinderVolumeDeleter) error
+}
+
+// cinderPersistentDisk volumes are disk resources provided by C3
```
Comment (Quentin-M) [inline pkg/volume/cinder_local/cinder.go]: Changed.

File: pkg/volume/cinder_local/cinder.go [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+	// Mounter interface that provides system calls to mount the global path to the pod local path.
+	mounter mount.Interface
+	plugin  *cinderLocalPlugin
+	volume.MetricsNil
+}
+
+type cinderVolumeMounter struct {
+	*cinderVolume
+	readOnly bool
+	pod      *api.Pod
```
Comment (Quentin-M) [inline pkg/volume/cinder_local/cinder.go]: Changed.

File: pkg/volume/cinder_local/cinder_util.go (hunk 1 above)
Comment (Quentin-M) [inline pkg/volume/cinder_local/cinder_util.go]: Afraid it would only contain this function?

File: pkg/volume/cinder_local/cinder.go [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+		SupportsSELinux: true,
+	}
+}
+
+func (m *cinderVolumeMounter) SetUp(fsGroup *int64) error {
+	return m.SetUpAt(m.GetPath(), fsGroup)
+}
+
+// SetUp bind mounts to the volume path.
+func (m *cinderVolumeMounter) SetUpAt(dir string, fsGroup *int64) error {
```
Comment (rootfs) [inline pkg/volume/cinder_local/cinder.go]: Implement attach/detach interface and make it a attachable volume, see https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/cinder/attacher.go

Comment (eyakubovich) [inline pkg/volume/cinder_local/cinder.go]: AFAIK, `Attacher` and `Detacher` interface is not called on the kubelet but rather on the controller. When the implementation ends up calling cloud provider, that's exactly what one wants. In this case the attachment needs to happen on the target kubelet. It exactly the same situation as with current rbd volume driver.

Comment (rootfs) [inline pkg/volume/cinder_local/cinder.go]: @eyakubovich we are addressing this issue in rbd, see #33660

**Comment by khrisrichardson** [comment]
@Quentin-M having not read the code or knowing the variances between the rackspace and gophercloud libs, do you know whether gophercloud is susceptible to the Cinder device mapping issue, which @kiall is attempting to workaround with https://github.com/kubernetes/kubernetes/pull/33270/files?

**Comment by anguslees** [comment]
@khrisrichardson: That's an issue with the information cinder server provides (the "device" value is mostly lies and wishful thinking) - and not something the client can fix.  The same workarounds will be necessary with both gophercloud/gophercloud and rackspace/gophercloud.

**Comment by anguslees** [comment]
Looks fine to me from a casual read.  I'm disappointed we couldn't share more of the code with the existing cinder volume plugin.

In particular, it seems like the main difference vs existing cinder plugin is that the keystone auth credentials now come from the (user-specified) volume metadata rather than the (admin-specified) provider config.  If this is the ~only difference, could cinder_local be basically a thin wrapper around creating an openstack provider instance with different credentials, and then using the existing cinder driver to do the attach/detach?  (Or if there is a gap, is there something easy we can do to make that possible rather than maintaining two separate codebases?)

**Comment by rootfs** [comment]
@anguslees +1
I would love to have a single cinder volume that works with openstack provider and also directly with keystone. Having two cinder volumes is confusing.

Is there anything existing cinder volume cannot be done while local cinder can? I see the flexible keystone  config is interesting, but unless a node can join multiple openstack, how is it different from using cloud config? 

Can multi attach support be considered in cloudprovider? @kubernetes/sig-openstack

**Comment by feiskyer** [comment]
@Quentin-M Really nice idea of supporting cinder volumes without cloud provider.

> I would love to have a single cinder volume that works with openstack provider and also directly with keystone. Having two cinder volumes is confusing.

+1. We reuse one cinder volume for both cases in [hypernetes](https://github.com/hyperhq/hypernetes/). The main differences are

- Auth: I think it's ok to get them from secret.
- How to mount the cinder backend volumes: without cloud provider, plugins couldn't attach the volume by Nova.

Other logics are generally same with existing cinder volume.

**Comment by xsgordon** [comment]
@rootfs multi-attach isn't quite fully complete from an OpenStack perspective (still some work outstanding in Nova) but I'm not sure if that actually impacts our ability to use it for these use cases?

**Comment by rootfs** [comment]
@xsgordon thanks for clarification. I agree multi-attach is not a must-have for cinder volume. We've marked cinder as RWO.

**Comment by rootfs** [comment]
@feiskyer Thanks for the pointer. I looked at Cinder bare metal volume you developed for hypernetes. It is an interesting project. Still, I am not convinced if we can reasonably get all cinder volume drivers to run on bare metal machines. What is in hypernetes looks to me serves a bridge to extract backend volume info.

**Comment by Quentin-M** [comment]
> Looks fine to me from a casual read. I'm disappointed we couldn't share more of the code with the existing cinder volume plugin.

On it!

**Comment by saad-ali** [comment]
CC @kubernetes/sig-storage Anyone have the bandwidth to review this?

**Comment by Quentin-M** [comment]
@saad-ali Sorry not to having been clearer. I am refactoring this quite heavily to support Attacher/Detacher interfaces and integrate into the existing Cinder volume. The code is already [there](https://github.com/Quentin-M/kubernetes/commits/volume-cinder-local-2) but not yet tested. These commits should not be reviewed. I could get early feedbacks on the WIP branch though if desired, I planned to ask @rootfs after some initial testing. Marked as WIP. Thanks.

**Comment by saad-ali** [comment]
> @saad-ali Sorry not to having been clearer. I am refactoring this quite heavily to support Attacher/Detacher interfaces and integrate into the existing Cinder volume. The code is already there but not yet tested. These commits should not be reviewed. I could get early feedbacks on the WIP branch though if desired, I planned to ask @rootfs after some initial testing. Marked as WIP. Thanks.

ACK Carry on

**Comment by Quentin-M** [comment]
Migrate rackspace/gophercloud -> gophercloud/gophercloud https://github.com/kubernetes/kubernetes/pull/36344

**Comment by Quentin-M** [comment]
Could anyone review this? 🎱 See above about the changes and which files are actually new. Thanks!

File: pkg/api/types.go [hunk 6]
```diff
@@ -845,6 +845,9 @@ type CinderVolumeSource struct {
 	// the ReadOnly setting in VolumeMounts.
 	// +optional
 	ReadOnly bool `json:"readOnly,omitempty"`
+	// Optional: Used to configure the Cinder credentials instead of using the
+	// Cloud Provider.
+	SecretRef string `json:"secretRef"`
```
Comment (rootfs) [inline pkg/api/types.go]: if it is optionally, can it be a pointer?

**Comment by rootfs** [comment]
@Quentin-M for `Detach` and `Unmount` refactoring, can you split into yet another commit, see https://github.com/kubernetes/kubernetes/pull/33660#issuecomment-258345434? 

+ @saad-ali

File: pkg/api/types.go (hunk 6 above)
Comment (Quentin-M) [inline pkg/api/types.go]: No other optional string fields are defined as pointers in _types.go_. Could you elaborate?

**Comment by Quentin-M** [comment]
@rootfs This is a complete rewrite of the work from my former coworker, which integrates the effort into the existing Cinder plugin rather than creating a new one and that changes dramatically the way it works in order to use the Attach/Detach pattern. Pretty much everything got modified (starting with the removal of the whole cinderLocal folder). Therefore I unfortunately don't clearly see how could I remove easily split the Detach/Unmount out.

@rootfs @saad-ali Regarding to https://github.com/kubernetes/kubernetes/pull/33660#issuecomment-258345434, the way it was before using the Attachable interface is that the plugin would persist on disk the secret behind the mount, and recover it right after unmount, for the detach operation. Now that Detach task happens separately from the Unmount, on potentially another machine, this trick can't be used anymore and we do still need the secret for Detach. If you can think of something else, I'd be glad to work around it.

Thank you for your time.

**Comment by rootfs** [comment]
ping @anguslees any take on gophercloud migration and code sharing?

**Comment by DTadrzak** [comment]
Looks good to me!

**Comment by intlabs** [comment]
Hey, Having looked through this it looks good, but would it be possible to add a CA Cert option to the keystone creds (and ideally a client cert option) as without this we would not be able to use for our deployments?

**Comment by Quentin-M** [comment]
@intlabs I just added support for both CA Cert and Client Certificate in a new commit. As discussed offline, it is assumed that the TLS configuration used for both the Keystone API and Cinder API are the same. Furthermore, if the detected Cinder endpoint doesn't support TLS at all, as you noted is a situation that still exists a bit out there, the TLS configuration is simply ignored for it.

File: pkg/api/types.go (hunk 6 above)
Comment (Quentin-M) [inline pkg/api/types.go]: @anguslees Great catch, thank you. I just updated it.

**Comment by PiotrProkop** [comment]
This PR is introducing a functionality which is really helpful for anyone who wants to integrate OpenStack and Kubernetes. Can we merge it as it is ?

**Comment by feiskyer** [comment]
Cool. LGTM.

File: hack/update-codegen.sh [hunk 7]
```diff
@@ -70,7 +70,7 @@ pkg/api
 pkg/api/v1
 $(
   cd ${KUBE_ROOT}
-  find pkg/apis -name types.go | xargs dirname | sort
+  find pkg/apis -name types.go | xargs -I {} dirname {} | sort
```
Comment (rootfs) [inline hack/update-codegen.sh]: is this change needed for this pr?

Comment (Quentin-M) [inline hack/update-codegen.sh]: Good catch, thank you. It comes from https://github.com/kubernetes/kubernetes/pull/36164, removing.

**Comment by rootfs** [comment]
I like the direction. Getting into 1.6 makes sense to me. 

- @kubernetes/sig-openstack is it ok to vendor in gophercloud?
- @saad-ali @jingxu97 is it ok to add volumeSpec to detach?

**Comment by rootfs** [comment]
@Quentin-M can you double check your gophercloud import? I find unrelated pkgs (cdn, compute, etc).

**Comment by jsafrane** [comment]
My personal feeling is that 724 files in this PR are quite a lot, it would be probably the largest Kubernetes  dependency.

Also, does this Cinder outside of OpenStack support all Cinder drivers or just RBD? Will all drivers need some specific code in Kubernetes? That's would be unfortunate, Cinder should be the abstraction layer here and Kubernetes should use the same APIs for all of them.

**Comment by Quentin-M** [comment]
@rootfs @jsafrane Probably an issue in my godep usage. https://github.com/kubernetes/kubernetes/pull/36344 changes way less. I'll figure that out. Thanks for pointing that out.

@jsafrane Cinder is _designed to present storage resources to end users that can be consumed_. We query Keystone/Cinder for authentication, volume creation, deletion, information fetching and locking - which is common for every backend drivers. Then, it is our responsibility to mount the actual volume on the kubelet - just like every other volume plugin (e.g. the _isci_ plugin shells out to `iscsiadm`). In the case of RBD, we have to use the _rbd_ kernel module. Therefore, yes, there is a small amount of code required for each backend that people would like to use. However, it is behind an interface (See `pkg/volume/cinder/drivers/drivers.go`) and is minimalistic (`Attach`, `Detach`, `IsAttached`). I don't see a way around it: we don't run Nova Compute on the kubelet host, nor do we run any hypervisor so we can't use what's existing there. But yes, Cinder is an abstraction and that's the whole point behind this PR: end-users can just specify their Keystone credentials, Volume ID and get going, instead of having to know every single driver-specific parameters and/or credentials.

**Comment by xsgordon** [comment]
In OpenStack-land this knowledge is not encoded directly in Cinder or Nova any more, but instead was moved out to the os-brick library (https://github.com/openstack/os-brick/tree/master/os_brick). That said since that:

1. It is a Python library and, like most shared libraries in OpenStack, nobody has ported them to Go as yet.
2. Nova and Cinder are the only current consumers of this I am aware of, so I wouldn't be shocked to find out that the model presented doesn't work for other consumers, even those that overcome the first point.

**Comment by Quentin-M** [comment]
@xsgordon Oh yeah, I entirely forgot about this library. I read about for the first time in the [e0ne's blog post](http://blog.e0ne.info/post/Attach-Cinder-Volume-to-the-Ironic-Instance-without-Nova.aspx) back then when we started thinking about this Kubernetes feature. Thanks for reminding me.

**Comment by Quentin-M** [comment]
@rootfs @jsafrane I just rebased and fixed the dependencies. We're back at something a bit more reasonable indeed. I vendored the same version that https://github.com/kubernetes/kubernetes/pull/36344 vendors to limit conflicts moving forward.

**Comment by Quentin-M** [comment]
Waiting for reviews before rebasing these generated files once again - it happens every 2 days.

**Comment by philips** [comment]
@saad-ali are you still the best person to review this or should it be reassigned to someone else?

**Comment by philips** [comment]
cc @barakmich @lpabon if you could help review too.

**Comment by Quentin-M** [comment]
Rebased.

**Comment by Quentin-M** [comment]
@saad-ali Hi. I hope you had great holidays. Could you please take a look? Several people have shown interest on this and it has received plenty reviews already too. Besides adding a new feature, it also clarifies the Cinder volume plugin code a lot.

**Comment by intlabs** [comment]
It would be great to see this merged if possible - currently, I'm relying on OS-Brick and flex-volume for bare-metal cinder PV's which is far from ideal, and having cinder as a first class citizen on bare-metal K8s would open up a whole load of use cases.

**Comment by Quentin-M** [comment]
Rebased. 

Does someone have some time to take a look at this? It is becoming more important as Cinder v1 will be gone in a week (Mitaka release). 

The refactoring done in this PR simplifies the structure and segregates the code related to the interaction between the plugin and the Cloud Provider itself in a single file, behind an interface, thus making the plugin straight-forward to maintain / update. 

Note however that the introduced feature (using Cinder w/o Cloud Provider) does support Cinder v2 natively.

**Comment by rootfs** [comment]
@saad-ali @jingxu97 any comment on adding `volume.Spec` to `Detach()`
https://github.com/kubernetes/kubernetes/pull/33944/commits/e192898bb6bb0bb60b548203501bf166565a8372#diff-cc7d1470983e68fa94e66db11aa03eed

**Comment by rootfs** [comment]
@anguslees @xsgordon 

- What's the plan for Cinder V1 -> V2 transition (#36976 is now closed) ? 
- Are you ok that this PR uses Cinder v2 API [directly](https://github.com/Quentin-M/kubernetes/blob/e192898bb6bb0bb60b548203501bf166565a8372/pkg/volume/cinder/manager_local.go) in the volume plugin (vs. in cloud provider)?

**Comment by Quentin-M** [comment]
cc @calebamiles

**Comment by saad-ali** [comment]
I'll take a look

File: pkg/api/v1/generated.proto [hunk 8]
```diff
@@ -207,11 +207,15 @@ message CinderVolumeSource {
...
 
+  // Optional: Used to configure the Cinder credentials instead of using the
+  // Cloud Provider.
+  optional LocalObjectReference secretRef = 3;
+
   // Optional: Defaults to false (read/write). ReadOnly here will force
   // the ReadOnly setting in VolumeMounts.
   // More info: http://releases.k8s.io/HEAD/examples/mysql-cinder-pd/README.md
   // +optional
-  optional bool readOnly = 3;
```
Comment (thockin) [inline pkg/api/v1/generated.proto]: @smarterclayton - I'm surprised we have protbufs where this tag renumbering is acceptable.  I see it's the internal type, so not serialized - where are we using the proto form of internal types?

File: pkg/volume/util/operationexecutor/operation_generator.go [hunk 9]
```diff
@@ -276,7 +276,7 @@ func (og *operationGenerator) GenerateDetachVolumeFunc(
 			err = og.verifyVolumeIsSafeToDetach(volumeToDetach)
 		}
 		if err == nil {
-			err = volumeDetacher.Detach(volumeName, volumeToDetach.NodeName)
+			err = volumeDetacher.Detach(volumeToDetach.VolumeSpec, volumeName, volumeToDetach.NodeName)
```
Comment (thockin) [inline pkg/volume/util/operationexecutor/operation_generator.go]: I am out of touch with the volume code these days, but once upon a time it was impossible to guarantee that the spec was available at Detach time, and therefore the name had to carry enough information to do a detach without the spec.

In fact, this is a large part of the reason that Attacher and Detacher exist.

Has ground truth changed such that I *know* the spec is available at detach time?

Comment (saad-ali) [inline pkg/volume/util/operationexecutor/operation_generator.go]: That is still the case. Which is why we left detach with only two parameters `volumeName` and `nodeName`. The full `volumeSpec` passed in via `volumeToDetach` is either a carry-over from the attach call (kept in-memory), or a "best-effort" reconstruction based on what was read of disk, and therefore fields other than the `volumeName` may not be correct.

@Quentin-M why is this change necessary? As far as I can tell, it's not used anywhere in your PR?

```GO
func (cdc *cdManagerCloud) Detach(_ *volume.Spec, ...)
```

File: pkg/volume/cinder/manager_local.go [hunk 10]
```diff
@@ -0,0 +1,559 @@
...
+// node.
+func (cdl *cdManagerLocal) Attach(spec *volume.Spec, nodeName types.NodeName) (string, error) {
+	volumeSource, readOnly, err := getVolumeSource(spec)
+	if err != nil {
+		return "", err
+	}
+	globalPDPath := makeGlobalPDName(cdl.plugin.host, volumeSource.VolumeID)
+	glog.V(4).Infof("Cinder (%v): Attach %v", cdl.GetName(), volumeSource.VolumeID)
+
+	client, err := newCinderClientFromSecret(volumeSource.SecretRef, cdl.plugin.host)
```
Comment (saad-ali) [inline pkg/volume/cinder/manager_local.go]: Why not call `newCinderClientFromSecret(...)` in `getManager(...)` and stick the resulting `gophercloud.ServiceClient` in the `cdManagerLocal` struct, instead of creating it on every call?

File: pkg/volume/cinder/cinder.go [hunk 11]
```diff
@@ -17,60 +17,72 @@ limitations under the License.
...
+	name := cinderVolumePluginName
+	return cd.plugin.host.GetPodVolumeDir(cd.podUID, kstrings.EscapeQualifiedNameForDisk(name), cd.volName)
 }
 
 type cinderPlugin struct {
 	host volume.VolumeHost
 	// Guarding SetUp and TearDown operations
 	volumeLocks keymutex.KeyMutex
+	// For tests purposes, otherwise detected using host:
+	cinderProvider CinderProvider
```
Comment (saad-ali) [inline pkg/volume/cinder/cinder.go]: `cinderProvider` is specific to the cloud provider implementation. Why stick it in this common `cinderPlugin` struct? Wouldn't it make more sense in `cdManagerCloud`?

File: pkg/volume/cinder/attacher.go [hunk 12]
```diff
@@ -17,191 +17,86 @@ limitations under the License.
...
-	}
-
-	volumeID := volumeSource.VolumeID
-
-	instances, res := attacher.cinderProvider.Instances()
-	if !res {
-		return "", fmt.Errorf("failed to list openstack instances")
-	}
-	instanceid, err := instances.InstanceID(nodeName)
+	manager, err := getManager(attacher.plugin, spec, nil)
```
Comment (saad-ali) [inline pkg/volume/cinder/attacher.go]: Should this `manager` be cached in `cinderVolume`, which already has a `manager cinderManager` field? If not, should that field be removed?

**Review by saad-ali — CHANGES_REQUESTED** [review]

File: pkg/volume/util/operationexecutor/operation_generator.go (hunk 9 above)
Comment (saad-ali) [inline pkg/volume/util/operationexecutor/operation_generator.go]: Oh, I see, you're using it for your new local (cloud-free) cinder implementation. The cloud provider implementations get credentials through external mechanism. For the local implementation you rely on reading credentials from a secret. And in order to get `SecretRef *LocalObjectReference`, which is available in the spec, and the pod namespace you need the volume spec on the detach.

```GO
func (cdl *cdManagerLocal) Detach(spec *volume.Spec, deviceName string, nodeName types.NodeName) error {
```

The problem with this, as stated above, is that the full `volumeSpec` passed into `volumeToDetach` can't really be relied on. If the controller binary is restarted between attach and detach and the PV or pod w/the volume def is deleted in the meantime, then the spec information is lost. Meaning if this happens (or if you delete your secret before detach), your volume gets stuck attached and you need to go and manually detach it. Would you be ok with this scenario?

Comment (Quentin-M) [inline pkg/volume/util/operationexecutor/operation_generator.go]: You got the reason right!

Unless you have a better idea (?), I believe that this imperfection can be tolerated for now given the raw improvement that this PR provides. If this becomes an actual issue for any user, then we could brainstorm harder and follow up with a smarter solution in due time. What do you think?

File: pkg/volume/cinder/attacher.go (hunk 12 above)
Comment (Quentin-M) [inline pkg/volume/cinder/attacher.go]: The Deleter and the Provisioner caches the `manager` inside `cinderVolume` as their constructors have either `volume.Spec` or `volume.VolumeOptions`, letting us determine which manager should be used right away (their instances are volume specific). Unfortunately, the Attacher and Detacher only get `volume.Spec` during the actual method calls and we therefore need to determine the manager at this time (their instance are not constructed as volume specific - even though it might be the case in practice).

**Comment by Quentin-M** [comment]
Rebased. I can re-test that PR fully once it gets a little further in the review process. Thanks.

**Comment by e0ne** [comment]
I'm not expert in kubernetes and Go, but this patch looks good from the Cinder's point of view. I'll try to spent more time playing with it soon

**Comment by dims** [comment]
@Quentin-M : will you be revisiting this PR?

**Comment by xrl** [comment]
Does this work mean cinder requests can connect volumes to dedicated hardware?

**Comment by mkutsevol** [comment]
Hey @xrl! I'm not the author/expert here, but when implementing v2 support I've explored this one a bit. This work is for supporting cinder volumes _without_ a cloud provider per se. Like a standalone mode for cinder volumes.

**Comment by j-griffith** [comment]
@Quentin-M let me know if you think you'll be picking this back up, if not I'd be happy to volunteer to help out with it, assuming it's something the community feels ok about moving forward with.  Personally, I think it's valuable to have this as supported.

**Comment by Quentin-M** [comment]
@j-griffith Thank you for following-up. I do think this has lot of value as well - just like various other members of the community. I think that this is not getting traction tho due to the interface change, that might be unsafe for kubelet restarts, and because of the shift towards having the drivers outside of the repository. I might be able to give it some work if the community wants to move forward too.

**Comment by j-griffith** [comment]
@Quentin-M Awesome!  Something I started thinking about a little bit today (and it might prove to be a bad idea so floating it here for all the world to chime in);  Rather than try and modify the existing Cinder code and also make the changes you have here to the interfaces, what about creating an independent volume specific driver keeping in line with the existing volume provider interface?

Honestly there's some room for clean up in the Cinder code that exists here anyway (deprecate/remove V1 as well as some consistency things), and this might be a good opportunity to tackle some of that, and at the same time we could work out the standalone Cinder model without having to introduce any risk or heavy lifting on the existing cloud-provider related side of it.  Then work out changes to the openstack cloudprovider (whether it's in or out of tree) to build on/use the volume plugin.

In terms of attach, just wanted to note that there is the os-brick-extension in cinderclient now that should enable all of the supported connectors in Cinder/Nova for Localhost.  Granted right now it wouldn't be the most elegant (the client being a python lib) but it's certainly doable.  It might at least be an interim solution if people are interested.  That means you in response to some of the earlier questions you then have the ability to attach ALL of the supported Cinder transport types (iSCSI-CHAP, FC, LIO, ISER, RBD etc).  Long term there is some interest from folks to build a brick golang package that would make things much nicer but that could be a ways off.  Anyway, just some thoughts for anybody/everybody interested here.

**Comment by mtanino** [comment]
@j-griffith @Quentin-M 
Hi, I think this PR needs to be discussed at sig-storage meeting again. Next meeting will be on July 20(Thursdays at 16:00 UTC).

Feature for v1.8 is under planning phase currently. Feel free to put the PR on the section "PRs to discuss or that need attention" then attend to the meeting!

- meeting note google doc
https://docs.google.com/document/d/1-8KEG8AjAgKznS9NFm3qWqkGyCHmvU6HVl0sk5hwoAE/edit#

**Comment by fabiand** [comment]
I wonder if this issue can be closed, considering that https://github.com/kubernetes-incubator/external-storage/pull/368 landed, this allows to consume Cinder managed volumes, using the already by Kubernetes supported transport protocols.

**Comment by jianglingxia** [comment]
Hi,now k8s support cinder volume in Iaas sceneo,is there a plan let k8s support cinder volume in bare metal ?or how can use it?thanks for all of you and for you reply! @rootfs @feiskyer @dims @FengyunPan

**Comment by dims** [comment]
@jianglingxia yes, you can use the standalone cinder provisioner with bare metal - https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/using-cinder-standalone-provisioner.md
